---
title: Annotate SVs with functional information
output: github_document
---

```{r include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, fig.width=10, fig.height=10)
```

```{r}
library(dplyr)
library(sveval)
library(GenomicRanges)
library(ggplot2)
library(gridExtra)
library(knitr)
library(rtracklayer)
## list of graphs
ggp = list()
```

## Gene annotation 

```{r}
if(!file.exists('gencode.v35.annotation.gtf.gz')){
  download.file('ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_35/gencode.v35.annotation.gtf.gz', 'gencode.v35.annotation.gtf.gz')
}

genc = import('gencode.v35.annotation.gtf.gz')
```

## SVs genotyped in 2,504 samples from 1000 Genomes Project

```{r}
## SVs grouped by site ('svsite' and 'clique' columns)
kgp = read.table('svs.2504kgp.svsite80al.tsv.gz', as.is=TRUE, header=TRUE)

## stats for each SV locus
## use the most frequent allele (and then the largest) for ac/af/size
## also saves sum/max/min across all alleles
kgp.s = kgp %>% arrange(desc(af), desc(size)) %>%
  group_by(seqnames, svsite, type, clique) %>%
  summarize(start=start[1], end=end[1], ac=ac[1], af=af[1], size=size[1], .groups='drop') %>%
  filter(size>=50) %>% makeGRangesFromDataFrame(keep.extra.columns=TRUE)
```

## Coding SVs

```{r}
genc.pc.cds = subset(genc, type=='CDS' & gene_type=='protein_coding')
ol.cds = findOverlaps(kgp.s, genc.pc.cds) %>% as.data.frame %>%
  mutate(gene=genc.pc.cds$gene_name[subjectHits]) %>%
  group_by(queryHits) %>% summarize(gene=paste(unique(sort(gene)), collapse=';'))

kgp.cds = kgp.s[ol.cds$queryHits]
kgp.cds$gene = ol.cds$gene

length(kgp.cds)
```

## Example: population-specific and novel

```{r}
## population specific information
freq.all = read.table('2504kgp.svsite80al.superpopfreq.tsv.gz', as.is=TRUE, header=TRUE)
pop.spec = read.table('pops-freq-1kgp-med1.tsv', as.is=TRUE, header=TRUE)

## novel annotation
novel = read.table('svsite.2504kgp.novel.tsv', as.is=TRUE, header=TRUE)

ex.df = pop.spec %>% filter(svsite %in% kgp.cds$svsite,
                            svsite %in% novel$svsite, 
                            af.med<.05) %>%
  arrange(desc(abs(af-af.med))) 
```

```{r, results='asis'}
tmp = lapply(1:min(10, nrow(ex.df)), function(ii){
  subset(kgp.cds, svsite==ex.df$svsite[ii]) %>% as.data.frame %>%
    mutate(coord=paste0('[', seqnames, ':', start, '-', end,
                        '](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=',
                        seqnames, '%3A', start, '%2D', end, ')')) %>% 
    select(coord, svsite, type, size, af, gene) %>% 
    kable %>% cat(sep='\n')
  cat('\n\n')
  freq.all %>% filter(svsite==ex.df$svsite[ii]) %>% arrange(desc(af)) %>% kable %>% cat(sep='\n')
  cat('\n\n')
})
```
